pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go

```diff
@@ -0,0 +1,144 @@
+package clientmiddleware
+
+import (
+	"context"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/services/caching"
+	"github.com/grafana/grafana/pkg/services/contexthandler"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// NewCachingMiddleware creates a new plugins.ClientMiddleware that will
+// attempt to read and write query results to the cache
+func NewCachingMiddleware(cachingService caching.CachingService) plugins.ClientMiddleware {
+	log := log.New("caching_middleware")
+	if err := prometheus.Register(QueryRequestHistogram); err != nil {
+		log.Error("error registering prometheus collector 'QueryRequestHistogram'", "error", err)
+	}
+	if err := prometheus.Register(ResourceRequestHistogram); err != nil {
+		log.Error("error registering prometheus collector 'ResourceRequestHistogram'", "error", err)
+	}
+	return plugins.ClientMiddlewareFunc(func(next plugins.Client) plugins.Client {
+		return &CachingMiddleware{
+			next:    next,
+			caching: cachingService,
+			log:     log,
+		}
+	})
+}
+
+type CachingMiddleware struct {
+	next    plugins.Client
+	caching caching.CachingService
+	log     log.Logger
+}
+
+// QueryData receives a data request and attempts to access results already stored in the cache for that request.
+// If data is found, it will return it immediately. Otherwise, it will perform the queries as usual, then write the response to the cache.
+// If the cache service is implemented, we capture the request duration as a metric. The service is expected to write any response headers.
+func (m *CachingMiddleware) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
+	if req == nil {
+		return m.next.QueryData(ctx, req)
+	}
+
+	reqCtx := contexthandler.FromContext(ctx)
+	if reqCtx == nil {
+		return m.next.QueryData(ctx, req)
+	}
```

Comment:
nit; I wonder if it makes sense to forward the request to the data source in these cases?

Reply:
In the case of alerting etc there will be no request context available

Reply:
Thanks, wasn't sure if this could happen but I wanted to guard against it anyway. Good to know about alerting. I'll write a unit test around empty req contexts.

Reply:
Done on [3153299](https://github.com/grafana/grafana/pull/65616/commits/315329998a2713fbc39bd4201acfa1a307ca49ec)